standards/core/rln-contract.md

```diff
@@ -0,0 +1,199 @@
+category: Standards Track
+tags: [waku/core-protocol]
+editor: Sergei Tikhomirov <sergei@status.im>
+contributors:
+---
+
+## Abstract
+
+This specification describes the smart contract that governs RLN memberships, in particular:
+
+- the overall smart contract functionality;
+- contract parameters;
+- the values of those parameters for the initial deployment on a mainnet chain;
+- the way in which the parameters can be modified.
+
+## Background
+
+[Rate-Limiting Nullifiers](https://rate-limiting-nullifier.github.io/rln-docs/) (RLN) is a ZK-based rate-limiting technique used in Waku.
+Before sending messages, users must register in a smart contract.
+They then attach a ZK proof of valid membership to every message.
+Relaying nodes check message validity and drop invalid messages.
+
+A message is considered valid in terms of RLN if:
+- its sender has an active RLN membership; and
+- the sender hasn't exceeded their rate limit within the current epoch.
+
+RLN is managed by a smart contract that keeps track of active memberships.
+Users interact with the contract to register a new membership and
+to obtain Merkle proofs and the tree root,
+which is necessary for proof generation and verification.
+
+In testnet deployment, the RLN contract does not require any payment apart from gas costs.
+This document aims to make the RLN contract suitable for mainnet deployment.
+In particular, the contract would require users to put down a deposit.
+The aim of the deposit initially is purely to protect the system from denial-of-service attacks using bandwidth capping.
+It may also be explored as a revenue stream for Waku in later versions.
+
+## Semantics
+
+This section is not intended to describe the full contract functionality.
+We only focus on functions required for managing memberships, expiration, deposits, etc.
+The document might later evolve into a full-fledged contract specification.
+
+The RLN contract provides the following functionalities:
+- register a membership in a given tier;
+- extend a membership;
+- withdraw the deposit for an expired membership.
+
+For the _Contract Owner_, the contract provides the following additional functionality:
+- change the modifiable parameters (see parameter table below);
+- (TBD) freeze certain functionality (e.g., in case of an attack, stop new registrations, deposit withdrawals, etc).
+
+In the initial deployment:
+- there is no slashing - that is, a user's deposit is not taken away in case of misbehavior;
+- membership expiration includes a grace period;
+- a user can extend their membership without putting up any additional deposit.
+
+### Membership life-cycle
+
+To register a membership, a user locks up a deposit.
+Each membership has an expiration date.
+After a membership expires, it enters a grace period, during which the user can extend it.
+Extending a membership requires no additional deposit.
+After the grace period, an expired membership can be taken over by another user.
```

Comment:
This point indeed needs a better wording. "Taken over" here doesn't mean that the original user loses their deposit. The point is (and IMO this is more of an implementation detail) that the non-zero elements in the RLN tree can (and should) be reused to limit tree growth. Therefore, if possible, new membership should reuse tree "slots" of expired memberships instead of creating new ones. What happens with the deposit of the expired membership, if it hasn't been withdrawn already, is a separate question. I would suggest saving this information into some data structure in the smart contract, so that the original user can withdraw their deposit even after the tree slot of their expired membership has been overwritten.

Reply:
> I think this is unfair. It shouldn't be penalized for being unaware of the expiration date.

The whole premise of an expiration date is that it expires. I am not sure to understand what is unfair here.